devito/operator/operator.py

```diff
@@ -297,19 +294,97 @@ def _lower_exprs(cls, expressions, **kwargs):
         expressions = [j for i in expressions for j in i._flatten]
         expressions = [indexify(i) for i in expressions]
         expressions = cls._apply_substitutions(expressions, subs)
+
         expressions = cls._specialize_exprs(expressions)
 
         return expressions
 
+    # Compilation -- Cluster level
+
+    @classmethod
+    def _specialize_clusters(cls, clusters, **kwargs):
+        """
+        Backend hook for specialization at the Cluster level.
+        """
+        return clusters
+
+    @classmethod
+    @timed_pass
+    def _lower_clusters(cls, expressions, **kwargs):
+        """
+        Clusters lowering:
+
+            * Group expressions into Clusters;
+            * Optimize Clusters for performance (flops, data locality, ...);
+            * Introduce guards for conditional Clusters
+        """
+        clusters = clusterize(expressions, dse_mode=kwargs['dse'])
+
+        clusters = cls._specialize_clusters(clusters)
+
+        return clusters
+
+    # Compilation -- ScheduleTree level
+
+    @classmethod
+    def _specialize_stree(cls, stree, **kwargs):
+        """
+        Backend hook for specialization at the Schedule tree level.
+        """
+        return stree
+
     @classmethod
-    def _specialize_iet(cls, iet, **kwargs):
+    @timed_pass
+    def _lower_stree(cls, clusters, **kwargs):
         """
-        Transform the IET into a backend-specific representation, such as code
-        to be executed on a GPU or through a lower-level system (e.g., YASK).
+        Schedule tree lowering:
+
+            * Turn a sequence of Clusters into a ScheduleTree;
+            * Derive and attach metadata for distributed-memory parallelism;
+            * Derive sections for performance profiling
```

Comment:
Both Drive on `specialize`?

Reply:
No, they are compulsory in lower. 

`_specialize_stree` is currently a no-op